google-cloud-logging/src/main/java/com/google/cloud/logging/Instrumentation.java

```diff
@@ -0,0 +1,196 @@
+import com.google.protobuf.Value;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Instrumentation {
+  public static final String DIAGNOSTIC_INFO_KEY = "logging.googleapis.com/diagnostic";
+  public static final String INSTRUMENTATION_SOURCE_KEY = "instrumentation_source";
+  public static final String INSTRUMENTATION_NAME_KEY = "name";
+  public static final String INSTRUMENTATION_VERSION_KEY = "version";
+  public static final String JAVA_LIBRARY_NAME_PREFIX = "java";
+  public static final String DEFAULT_INSTRUMENTATION_VERSION = "UNKNOWN";
+  public static final int MAX_DIAGNOSTIC_VALUE_LENGTH = 14;
+  public static final int MAX_DIAGNOSTIC_ENTIES = 3;
+  private static boolean instrumentationAdded = false;
+
+  public static Tuple<Boolean, Iterable<LogEntry>> populateInstrumentationInfo(
+      Iterable<LogEntry> logEntries) {
+    boolean isWritten = setInstrumentationStatus(true);
+    if (isWritten) return Tuple.of(false, logEntries);
+    List<LogEntry> entries = new ArrayList<>();
+
+    for (LogEntry logEntry : logEntries) {
+      // Check if LogEntry has a proper payload and also contains a diagnostic entry
+      if (!isWritten
+          && logEntry.getPayload().getType() == Type.JSON
+          && logEntry
+              .<Payload.JsonPayload>getPayload()
+              .getData()
+              .containsFields(DIAGNOSTIC_INFO_KEY)) {
+        try {
+          ListValue infoList =
+              logEntry
+                  .<Payload.JsonPayload>getPayload()
+                  .getData()
+                  .getFieldsOrThrow(DIAGNOSTIC_INFO_KEY)
+                  .getStructValue()
+                  .getFieldsOrThrow(INSTRUMENTATION_SOURCE_KEY)
+                  .getListValue();
+          entries.add(createDiagnosticEntry(null, null, infoList));
+          isWritten = true;
+        } catch (Exception ex) {
+          System.err.println("ERROR: unexpected exception in populateInstrumentationInfo: " + ex);
+        }
+      } else {
+        entries.add(logEntry);
+      }
+    }
+    if (!isWritten) {
+      entries.add(createDiagnosticEntry(null, null, null));
+    }
+    return Tuple.of(true, entries);
+  }
+
+  /**
+   * The helper method to generate a log entry with diagnostic instrumentation data.
+   *
+   * @param libraryName {string} The name of the logging library to be reported. Should be prefixed
+   *     with 'java'. Will be truncated if longer than 14 characters.
+   * @param libraryVersion {string} The version of the logging library to be reported. Will be
+   *     truncated if longer than 14 characters.
+   * @returns {LogEntry} The entry with diagnostic instrumentation data.
+   */
+  public static LogEntry createDiagnosticEntry(String libraryName, String libraryVersion) {
+    return createDiagnosticEntry(libraryName, libraryVersion, null);
+  }
+
+  private static LogEntry createDiagnosticEntry(
+      String libraryName, String libraryVersion, ListValue existingLibraryList) {
```

Comment:
nit: the code that creates protobuf entry is hard to read. would it be possible to create the `Dictionary<?,?>` based payload? i think the code will be more readable and further developers can easily guess the instrumentation payload structure.

Reply:
Sure - added a separate `Object` to be used for synchronization

Reply:
> I did not find instrumentation for the calls to `Logging.Write()` and `Logging.WriteAsync()`. Are they expected to be in a separate PR? I recommend to instrument all paths before merging.

Right - good catch! Eventually `Logging.writeAsync()` is private, but `Logging.write()` is public and I missed it

Reply:
> nit: the code that creates protobuf entry is hard to read. would it be possible to create the `Dictionary<?,?>` based payload? i think the code will be more readable and further developers can easily guess the instrumentation payload structure.

To be honest, I prefer the protobuf structure since it is safer from types handling perspective. If you look into code I added, it uses bunch of useful type checking methods like` hasStructValue()` or strongly types setters/getters like `setStructValue()` /  `getValuesList()`. Also I liked methods as `getFieldsOrThrow()` which pretty convenient. Of course I could achieve type casting of `Map<String, Object>` values, but it would make code to be complex in other parts like reading...